contracts/src/v0.8/automation/upkeeps/LinkAvailableBalanceMonitor.sol

```diff
@@ -163,27 +154,68 @@ contract LinkAvailableBalanceMonitor is ConfirmedOwner, Pausable, AutomationComp
     return targetsToFund;
   }
 
-  /// @notice Send funds to the targets provided.
-  /// @param targetAddresses the list of targets to fund
   function topUp(address[] memory targetAddresses) public whenNotPaused {
-    uint256 topUpAmount = s_topUpAmount;
-    uint256 stopIdx = targetAddresses.length;
-    uint256 numCanFund = LINK_TOKEN.balanceOf(address(this)) / topUpAmount;
-    stopIdx = numCanFund < stopIdx ? numCanFund : stopIdx;
-    for (uint256 idx = 0; idx < stopIdx; idx++) {
-      (bool exists, uint256 minBalance) = s_watchList.tryGet(targetAddresses[idx]);
-      if (!exists) {
-        emit TopUpBlocked(targetAddresses[idx]);
-        continue;
+    MonitoredAddress memory target;
+    uint256 localBalance = LINK_TOKEN.balanceOf(address(this));
+    address[] memory targetsTopUp = new address[](targetAddresses.length);
+    for (uint256 idx = 0; idx < targetAddresses.length; idx++) {
+      address targetAddress = targetAddresses[idx];
+      target = s_targets[targetAddress];
+      if (localBalance >= target.topUpAmount && _needsFunding(targetAddress, target.minBalance)) {
+        targetsTopUp[idx] = targetAddress;
+        localBalance -= target.topUpAmount;
+      } else {
+        emit TopUpBlocked(targetAddress);
       }
-      (bool needsFunding, address target) = _needsFunding(targetAddresses[idx], minBalance);
-      if (!needsFunding) {
-        emit TopUpBlocked(targetAddresses[idx]);
+    }
+    for (uint256 idx = 0; idx < targetsTopUp.length; idx++) {
+      address targetAddress = targetsTopUp[idx];
+      if (targetAddress == address(0)) {
         continue;
       }
-      LINK_TOKEN.transfer(target, topUpAmount);
-      emit TopUpSucceeded(targetAddresses[idx]);
+      target = s_targets[targetAddress];
+      bool success = LINK_TOKEN.transfer(targetAddress, target.topUpAmount);
+      if (success) {
+        target.lastTopUpTimestamp = uint56(block.timestamp);
+        emit TopUpSucceeded(targetAddress);
+      } else {
+        emit TopUpFailed(targetAddress);
+      }
+    }
```

Comment:
is it possible to combine these for loops? I feel like there is a fair amount of bloat we could cut. Ex...
* `targetAddresses` and `targetsTopUp` are essentially duplicate arrays
* we check `if (targetAddress == address(0))` twice - once in `_needsFunding()` and then again in the 2nd for loop (because the 2nd loop potentially contains zero addresses)

I think doing this in one for loop would kind of break the "checks..effects..interactions" pattern, but I don't think we are vulnerable to reentrancy because we are only doing a `transfer()` call on the link token (safeTransferFrom might make us vulnerable though :sweat_smile:). But we could always just add a reentrancy guard if that is a concern.